### ISSUE TITLE
False positive: button-name rule fails for button labelled using label/for

### ISSUE THREAD

**Issue by iamrafan** [dequelabs/axe-core#1393]
- Link to demo: https://codepen.io/iamrafan/pen/RdNEgV?editors=1000#0

- AT Testing:

  | AT/Browser | Announced as
-- | -- | --
1 | Edge/Narrator | 'Enabled and   Checked' toggle switch
2 | Edge/JAWS | 'Enabled and   checked' Toggle button Pressed
3 | Edge/NVDA | toggle button   pressed  'Enabled and checked'
4 | Chrome/NVDA | toggle button  pressed    checked   'Enabled and   checked'
5 | Chrome/JAWS | 'Enabled and   checked' Switch Pressed
6 | IE/JAWS | 'Enabled and   checked' Switch On
7 | IE/NVDA | Button 'Enabled   and checked'
8 | Firefox/NVDA | toggle button not   pressed  checked   'Enabled and   checked'

<pre><code>axe-core version: 3.1.2
axe-Coconut version: 3.7.0</code></pre>

**Comment by WilcoFiers** [comment]
I had no idea that works. Good find Guru.

**Comment by straker** [comment]
I did some additional testing using the `<button>` like elements we use for the `button-name` check. Here's the results https://codepen.io/straker/pen/WmqGmK (someone is testing Dragon for me tomorrow).

For the most part, it looks like everything but `<div role="button">` is supported in everything but IE 11 and JAWS, where the support is almost flipped (surprisingly it supports `<div role="button">`).

**Comment by straker** [comment]
Update from the testing with Dragon:

> "Okay, so… Dragon is a fickle beast and its always hard to tell when it can understand what you are saying versus not actually supporting a certain kind of coding pattern. However, I’m now confident that it doesn’t understand this pattern at all (which makes sense because it doesn’t use the underlaying accessibility APIs)
>
> I had to hack your example a bit to be able to test it. Also, codepen was down so I moved the example to glitch. I added an alert so that I was aware which button was activated. I also added a plain old normal button as a control group if you will :stuck_out_tongue:. What made me confident that Dragon doesn’t support this pattern is that I could say ‘click button’ and it would flag all buttons and let me pick which one to select. But if I said ‘click my label’ or ‘click label’, only the control group button would be flagged by button (even though all labels contain said text). I was unable to activate any of the buttons by saying ‘click my button 1’, etc.
>
> https://chalk-agenda.glitch.me/

**Comment by straker** [comment]
After doing lots of testing, it seems JAWS does not really support this technique. It only supported it when `<button role="switch">` or when the button has aria states (for some reason). Combined with the fact that this technique was met with "Wait, that really works?" when discussing it means most people probably don't expect it to work. 

Due to these issues, and that Dragon does not support this technique at all to compute the accessible name, I propose that we do not support this technique.

@dylanb

**Comment by dylanb** [comment]
I agree
